src/vdbe.c

```diff
@@ -8072,6 +8072,40 @@ case OP_VInitIn: {        /* out2 */
 }
 #endif /* SQLITE_OMIT_VIRTUALTABLE */
 
+#ifndef SQLITE_OMIT_VIRTUALTABLE
+/* Opcode: VPrepareSql P1 * * * *
+**
+** P1 is a cursor opened using VOpen.
+**
+** This opcode invokes the xPrepareSql method on the virtual table specified
+** by P1. The SQL text parameter to xPrepareSql is obtained from Vdbe* `p`.
+**
+*/
+case OP_VPrepareSql: {
+  const sqlite3_module *pModule;
+  sqlite3_vtab_cursor *pVCur;
+  sqlite3_vtab *pVtab;
+  VdbeCursor *pCur;
+
+  pCur = p->apCsr[pOp->p1];
+  assert( pCur!=0 );
+  assert( pCur->eCurType==CURTYPE_VTAB );
+  pVCur = pCur->uc.pVCur;
+  pVtab = pVCur->pVtab;
+  pModule = pVtab->pModule;
+
+  /* Invoke the xPrepareSql method */
+   if( pModule->iVersion>=4 ){
+    if( pModule->xPrepareSql && p->zSql ){
```

Comment:
Question - in which cases is `p->zSql` null? I wonder if the responsibility for checking and handling NULL shouldn't be pushed to the implementation of `xPrepareSql` - perhaps somebody would like to do something specific even for cursors that don't have the SQL text?

This is a genuine question, as I haven't implemented any virtual SQLite tables in my life (: if it doesn't make sense, then feel free to leave the code as is.

Reply:
I have added that as a sanity check. A NULL value for `p->zsql` would be possible when handling the `Init` or `Trace` opcodes, but that is not expected when processing a valid SQL statement.